opm/parser/eclipse/EclipseState/Schedule/SegmentSet.cpp

```diff
@@ -0,0 +1,277 @@
+        return m_segments;
+    }
+
+    SegmentPtr& SegmentSet::operator[](size_t idx) {
+        return m_segments[idx];
+    }
+
+    SegmentPtr SegmentSet::operator[](size_t idx) const {
+        return m_segments[idx];
+    }
+
+    bool SegmentSet::numberToLocation(const int segment_number, int& location) const {
+         // std::map<int, int>::iterator it;
+         auto it = m_number_to_location.find(segment_number);
+         if (it != m_number_to_location.end()) {
+             location = it->second;
+             return true;
+         } else {
+             return false;
+         }
+    }
+
+    int SegmentSet::numberToLocation(const int segment_number) const {
+        auto it = m_number_to_location.find(segment_number);
+        if (it != m_number_to_location.end()) {
+            return it->second;
+        } else {
+            return -1;
+        }
+    }
+
+    void SegmentSet::addSegment(SegmentPtr new_segment) {
+        // decide whether to push_back or insert
+        int segment_number = new_segment->segmentNumber();
+
+        int segment_location; // storage location
+        bool find = numberToLocation(segment_number, segment_location);
+
+        if (!find) { // it is a new segment
+            m_segments.push_back(new_segment);
+            m_number_to_location[segment_number] = m_number_segment;
+            ++ m_number_segment;
+        } else { // the segment already exists
+            m_segments[segment_location] = new_segment;
+        }
+    }
+
+    void SegmentSet::addSegmentINC(SegmentPtr /* new_segment */) {
+        throw std::logic_error("SegmentSet::addSegmentINC() not implemented yet");
+    }
+
+    void SegmentSet::addSegmentABS(SegmentPtr /* new_segment */) {
+        throw std::logic_error("SegmentSet::addSegmentABS() not implemented yet");
+    }
+
+    SegmentSet* SegmentSet::shallowCopy() const {
+        SegmentSet* copy = new SegmentSet();
+        copy->m_well_name = m_well_name;
+        copy->m_number_branch = m_number_branch;
+        copy->m_number_segment = m_number_segment;
+        copy->m_depth_top = m_depth_top;
+        copy->m_length_top = m_length_top;
+        copy->m_volume_top = m_volume_top;
+        copy->m_length_depth_type = m_length_depth_type;
+        copy->m_comp_pressure_drop = m_comp_pressure_drop;
+        copy->m_multiphase_model = m_multiphase_model;
+        copy->m_x_top = m_x_top;
+        copy->m_y_top = m_y_top;
+        copy->m_number_to_location = m_number_to_location;
+        copy->m_segments.resize(m_segments.size());
+        for (size_t i = 0; i < m_segments.size(); ++i) {
+            copy->m_segments[i] = m_segments[i];
+        }
+        return copy;
+    }
+
+    void SegmentSet::segmentsFromWELSEGSKeyword(DeckKeywordConstPtr welsegsKeyword, DeckKeywordConstPtr nsegmxKeyword) {
+        // get the required information for multi-segment wells
+        const int nsegmx = nsegmxKeyword->getRecord(0)->getItem("NSEGMX")->getInt(0);
```

Comment:
Are these variables just Eclipse style memory management, or do we actually need them for semantic correctness?